### PR TITLE
Prevents unnecessary calculation of stats

### LINKED ISSUE #4895: HazelCast UI issues in 3.x versions
We have observed UI issues in hazelcast in 3.x versions.

UI starts wobbling with increase in entries.

Total maps : 5 

Total entries per Map is around 2 Million except one map which has around 93 thousand entries.  

Map Sizes:

Maps having 2 Millions entries:
Map1 - 375 MB 
Map2- 935 MB
Map 3- 880 MB
Map 4- 415 Mb

Map with 93 k entries - 16 MB

We have tried hazelcast 3.3.3 , 3.3.5 and 3.4.1 . It works fine in hazelcast 2.6.9 version.

### PR DESCRIPTION
**Pull request by emrahkocaman**
fixes #4895, Stats were being calculated twice just to collect data structure names.
Also `StateSendThread` sleep time is now calculated depending on the elapsed time on stat gathering.

### CODE REVIEW

**Comment by kobalski** [comment]
:+1:

**Comment by mesutcelik** [comment]
is it possible to add some tests?

**Comment by pveentjer** [comment]
@emrahkocaman can you add some tests? Then we can merge this pr.

**Comment by muratayan** [comment]
@mesutcelik @pveentjer We spoke with @emrahkocaman, I will work on the tests

File: hazelcast/src/test/java/com/hazelcast/internal/management/ManagementCenterServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.URL;
+
+import static java.lang.String.format;
+import static org.junit.Assert.assertNotEquals;
+
+
+public class ManagementCenterServiceTest extends HazelcastTestSupport {
```
Comment (bilalyasar) [inline hazelcast/src/test/java/com/hazelcast/internal/management/ManagementCenterServiceTest.java]: can you add category?

File: hazelcast/src/test/java/com/hazelcast/internal/management/ManagementCenterServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    @Before
+    public void setUp() throws Exception {
+        URL root = new URL(MancenterServlet.class.getResource("/"), "../test-classes");
+        String baseDir = new File(root.getFile().replaceAll("%20", " ")).toString();
+        String sourceDir = baseDir + "/../../src/test/webapp";
+        String sourceName = "server_config.xml";
+        portNum = availablePort();
+        jettyServer = new JettyServer(portNum, sourceDir, sourceName);
+
+        hazelcastInstance = Hazelcast.newHazelcastInstance(makeConfig());
```
Comment (muratayan) [inline hazelcast/src/test/java/com/hazelcast/internal/management/ManagementCenterServiceTest.java]: This should be shut down afterwards, right?

File: hazelcast/src/test/java/com/hazelcast/internal/management/JettyServer.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package com.hazelcast.internal.management;
+
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.nio.SelectChannelConnector;
+import org.eclipse.jetty.webapp.WebAppContext;
+
+public class JettyServer implements ServletContainer {
```
Comment (bilalyasar) [inline hazelcast/src/test/java/com/hazelcast/internal/management/JettyServer.java]: i guess, no need to implement ServletContainer

File: hazelcast/src/test/java/com/hazelcast/internal/management/ManagementCenterServiceTest.java (hunk 2 above)
Comment (bilalyasar) [inline hazelcast/src/test/java/com/hazelcast/internal/management/ManagementCenterServiceTest.java]: definitely

File: hazelcast/src/test/java/com/hazelcast/internal/management/JettyServer.java (hunk 3 above)
Comment (emrahkocaman) [inline hazelcast/src/test/java/com/hazelcast/internal/management/JettyServer.java]: :+1: , Can you please delete `ServletContainer` class?

**Comment by emrahkocaman** [comment]
Thanks for the tests @muratayan .  @mesutcelik I think this pr can be merged now.

File: hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java [hunk 4]
```diff
@@ -295,8 +296,10 @@ private long calcUpdateInterval() {
         public void run() {
             try {
                 while (isRunning()) {
+                    long start = Clock.currentTimeMillis();
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java]: Always include the unit. E.g. startMs.

File: hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java [hunk 5]
```diff
@@ -307,8 +310,11 @@ public void run() {
             }
         }
 
-        private void sleep() throws InterruptedException {
-            Thread.sleep(updateIntervalMs);
+        private void sleepIfPossible(long elapsed) throws InterruptedException {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java]: unit, e.g. elapsedMs.

File: hazelcast/src/test/java/com/hazelcast/internal/management/ManagementCenterServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.URL;
+
+import static java.lang.String.format;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+
+@Category(QuickTest.class)
+public class ManagementCenterServiceTest extends HazelcastTestSupport {
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/internal/management/ManagementCenterServiceTest.java]: I don't think we should classify this as a 'quick' test. This looks like an integration test to me. It certainly isn't a unit test. 

How much time does it take to run?

**Comment by pveentjer** [comment]
Also squash the commits.

**Pull request merged**